packages/pkl.experimental.structuredRead/structuredRead.pkl

```diff
@@ -0,0 +1,276 @@
+
+/// Function used to read in resources. Can be overwritten to change how
+/// values are read. Useful for testing.
+hidden readFunc = (s) -> read?(s)
+
+function applyToClass(target: Class, path: List<String>): Mapping<String, Any?> =
+  let (reflected = reflect.Class(target))
+  new Mapping {
+    for (name, property in reflected.properties) {
+      when (property.type is Coerceable) {
+        [name] = let (potentialValue = readValueToTarget(path, name, property))
+            if (!(potentialValue is ReadFailure)) potentialValue
+            else if (property.type is reflect.NullableType) property.defaultValue
+            else property.defaultValue ?? potentialValue
+      }
+    }
+  }
+
+typealias Coerceable = reflect.DeclaredType | reflect.NullableType | reflect.TypeAlias | reflect.UnknownType
+
+local function assertNoFailures(target: Class, output: Mapping<String, Any?>|Dynamic): Boolean =
+  let (fails = findFailures(output))
+    fails.isEmpty ||
+      let (msg = new Listing {
+        "Error processing \(if (fails.length > 1) "properties" else "property") in \(target):\n"
+        local maxPathLength = fails.map((f) -> f.path.join(pathSeparator).length).max
+        for (fail in fails) {
+          let (path = fail.path.join(pathSeparator))
+            "\t'\(path)'\(" ".repeat(maxPathLength - path.length)): \(fail.message)"
+        }
+      }.join("\n"))
+        throw(msg)
+
+local function longestPath(fails: List<Failure>) =
+  fails.map((f) -> f.path.join(pathSeparator).length).max
+
+local function findFailures(output: Mapping<String, Any?>|Dynamic): List<Failure> =
+  output.toMap().fold(List(), (acc: List<Failure>, _, value) ->
+    if (value is Failure)
+      acc.add(value)
+    else if (value is Mapping)
+      acc + findFailures(value)
+    else if (value is Dynamic)
+      acc + findFailures(value)
+    else
+      acc
+  )
+
+abstract class Failure {
+  message: String
+  path: List<String>
+}
+
+class ReadFailure extends Failure
+class CoerceFailure extends Failure
+
+local function ReadFail(path: List<String>, msg: String) = let (_path = path ) new ReadFailure {
+  message = msg
+  path = _path
+}
+
+local function CoerceFail(path: List<String>, msg: String) = let (_path = path ) new CoerceFailure {
+  message = msg
+  path = _path
+}
+
+local function parseArraySyntax(input: String): Listing = new {
+  when (input.startsWith("{") && input.endsWith("}")) {
+    ...input.drop(1).dropLast(1).split(",").map((e) -> e.trim())
+  } else {
+    input
+  }
+}
+
+local function applyToDynamic(path, default: Dynamic) =
+  if (default.toList().isEmpty && default.toMap().isEmpty)
+    // Do some special handling for empty dynamic types
+     let (potentialValue = readValue(path))
+       if (potentialValue is ReadFailure)
+         default
+       else if (potentialValue is String)
+         parseArraySyntax(potentialValue)
+       else
+         potentialValue ?? default
```

Comment:
Nit; indent

Also "empty dynamic types" is a bit of a contradiction in terms. There is only one `Dynamic` type. If you're talking about that one, the comment should be `for empty Dynamic objects.`